Refactor opaque dtype implementation.

jakevdp opened the pull request:
This makes it closer to numpy, with `dtypes.OpaqueDtype` analogous to `np.dtype`, and `dtypes.opaque` analogous to `np.numeric`. This will let us replace the `dtypes.is_opaque_dtype(dtype)` function with `jnp.issubdtype(dtype, dtypes.opaque)`.

This underlying change will make the goal in #16781 much easier to realize.

froystig commented on jax/_src/dtypes.py:
```diff
@@ -45,11 +45,32 @@
+  def type(self) -> type: ...
+
 
 def is_opaque_dtype(dtype: Any) -> bool:
```
```suggestion
# TODO(vanderplas,frostig): remove in favor of inlining `issubdtype`
def is_opaque_dtype(dtype: Any) -> bool:
```

froystig reviewed (approved):
We should also eventually clean up `DType` versus `OpaqueDType` annotations, maybe by having `BasicDType`, `OpaqueDType`, and `DType` as the union of the two.

jakevdp committed:
Refactor opaque dtype implementation.

This makes it closer to numpy, with dtypes.OpaqueDtype analogous to np.dtype,
and dtypes.opaque analogous to np.numeric. This will let us replace the
dtypes.is_opaque_dtype function with jnp.issubdtype(dtype, dtypes.opaque).
